prog.go

```diff
@@ -400,7 +408,7 @@ func newProgramFromFD(fd *sys.FD) (*Program, error) {
 		return nil, fmt.Errorf("discover program type: %w", err)
 	}
 
-	return &Program{"", fd, "", "", info.Type}, nil
+	return &Program{"", fd, "", "", info.Type, AttachNone}, nil
```

Comment:
This is going to be a problem. It means that some programs can't be attached, depending on how the program was created. For example pinned programs would fail due to this.

You either need to find a reliable way to populate this field or change the API.

(I haven't had the time to look at the rest of the PR, sorry.)

Reply:
hi @lmb , I just picked this up again and I have to choose between:

1. deprecate `AttachTracing` and introduce `AttachFEntry` , `AttachFExit`, `AttachModRet`

2. add `AttachType ebpf.AttachType` to TracingOptions and let the user select the correct attach type

I wonder what is the lesser evil between deprecating an api or adding a redundant option.

My preference goes to `1.`, since imho it will be more clear to users long term and `2.` requires one or two more input checks, but let me know if you think `2.` is better

Reply:
> I wonder what is the lesser evil between deprecating an api or adding a redundant option.

I prefer option 2. Introducing separate functions makes it hard to write code that is "attachment agnostic": you  need to pass around a function pointer or pass around an extra AttachType anyways. If we stick the AttachType into options we may be able to make it optional in the future, if a way to retrieve the type from the kernel comes around.